### ISSUE TITLE
Regression on rest_framework : Value 'serializer.data' is unsubscriptable

### ISSUE THREAD

**Issue by romainletendart** [PyCQA/pylint#4694]
`rest_framework.serializers.Serializer.data` is a property that returns a `ReturnDict` object. That class is a subclass of the well known `collections.OrderedDict` and is obviously subscriptable. What I notice is that the error only appears when the field we add to the serializer gets the very same name 'data'.

What might get pylint confused is that there is some metaclass magic going on when declaring a Serializer subclass. If we declare a 'data' field/attribute, it won't in the end override the original 'Serializer.data' attribute but end up within that very attribute.

Pylint 2.8.3 didn't raise any errors for this case.

### Steps to reproduce

Given a file `serializer_data_unsubscriptable.py`:
```python
import django
from django.conf import settings
from rest_framework import serializers


class DataSerializerWithError(serializers.Serializer):
    data = serializers.JSONField()
    def update(self, instance, validated_data):
        pass

    def create(self, validated_data):
        pass


class DataSerializerWithoutError(serializers.Serializer):
    payload = serializers.JSONField()
    def update(self, instance, validated_data):
        pass

    def create(self, validated_data):
        pass


def print_data():
    settings.configure()
    django.setup()

    serializer = DataSerializerWithError(data={'data': 42})
    if not serializer.is_valid():
        print("Failure!")
        return
    print(serializer.data['data'])

    serializer = DataSerializerWithoutError(data={'payload': 42})
    if not serializer.is_valid():
        print("Failure!")
        return
    print(serializer.data['payload'])
```

### Current behavior

Result of `pylint serializer_data_unsubscriptable.py`:
```
$ pylint serializer_data_unsubscriptable.py
************* Module serializer_data_unsubscriptable
serializer_data_unsubscriptable.py:32: [E1136(unsubscriptable-object), print_data] Value 'serializer.data' is unsubscriptable
```

### Expected behavior

No errors.

### pylint --version output

Result of `pylint --version` output:

```
pylint 2.9.3
astroid 2.6.2
Python 3.8.9 (default, Apr 19 2021, 08:40:21) 
[GCC 10.2.0]
```

**Comment by LefterisJP** [comment]
I also started getting a few false-positive unsuscriptable objects in our projects with 2.9.3. I suspect underlying reason is the same. Have not managed to create a self-contained example yet.

**Comment by timmartin** [comment]
It looks like this has been caused by a change in Astroid, rather than Pylint. If I use Pylint v2.8.3 but upgrade Astroid to v2.6.6, I get the warning.

The direct cause of this is that in `pylint/checkers/typecheck.py`, in `TypeCheckers.visit_subscript()`, the older Astroid is failing to infer the type of `serializer.data` so the checker abandons the check. The more recent Astroid successfully infers the type of `serializer.data` and therefore the check takes place.

Is this actually a false positive? In the case of `DataSerializerWithoutError` the member `serializer.data` corresponds to the `data` property on the `Serializer` base class, which is subscriptable. In the case of `DataSerializerWithError` you're declaring `data` as a `JSONField` in the derived class, which will hide the subscriptable member on the base class. It looks like the `JSONField` isn't subscriptable. I'd expect there to be subscriptable data available as `serializer.validated_data['data']` which will include the deserialized JSON object.

Is there some additional DRF magic here I'm forgetting about? Even if there is, I think this is a case where Pylint has never had the necessary logic to infer this, and it's not a regression as such.

**Comment by romainletendart** [comment]
@timmartin Creating a `data` attribute in a `Serializer` subclass just seems to work when the code is run through a Python 3.8.11 interpreter. We do get `42` printed twice. The magic happens on those lines: https://github.com/encode/django-rest-framework/blob/cdd53c7de912d5868c96f4e3883df248a3e6341d/rest_framework/serializers.py#L533-L549.

I agree with you however on the fact that having an overriden `data` attribute brings confusion.

**Comment by timmartin** [comment]
It took me a while, but I think I've figured out what's going on here. Not surprisingly, it's down to the metaclass:

    class SerializerMetaclass(type):    
        @classmethod
        def _get_declared_fields(cls, bases, attrs):
            fields = [(field_name, attrs.pop(field_name))
                      for field_name, obj in list(attrs.items())
                      if isinstance(obj, Field)]
            fields.sort(key=lambda x: x[1]._creation_counter)
            # ....

        def __new__(cls, name, bases, attrs):
            attrs['_declared_fields'] = cls._get_declared_fields(bases, attrs)
            return super().__new__(cls, name, bases, attrs)

Within `__new__`, the `attrs` dict has any values stripped out of it if they are instances of `Field`, meaning that in our case the `data` property in the base class will be visible because the derived class doesn't this as an attribute.

As an aside, it was a real pain to figure this out because there's a method called `_get_declared_fields` that modifies its arguments; I wasted a lot of time having made the assumption that this wouldn't be changing the `attrs` value.

I think the right way to fix this will be to add some logic to the Astroid brain so that it can correctly infer that these attributes won't be present.

**Comment by clavedeluna** [comment]
This is reproducible with first running `pip install djangorestframework`
